README.md
# Comment Link Moderation Whitelist

A plugin for WordPress that allows for whitelisted URLs to not count against the comment moderation max number of links limit.

This plugin is available in the WordPress Plugin Directory: https://wordpress.org/plugins/comment-link-moderation-whitelist/


## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress
3. Via the 'Settings' -> 'Discussion' admin page, define some URLs in the 'Comment Link Moderation Whitelist' field.


## Additional Documentation

See [readme.txt](https://github.com/coffee2code/comment-link-moderation-whitelist/blob/master/readme.txt) for additional usage information. See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/comment-link-moderation-whitelist/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/comment-link-moderation-whitelist/).

If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/comment-link-moderation-whitelist/reviews/) for it in the WordPress Plugin Directory or considerating a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# Changelog

## 1.1.3 _(2020-05-12)_
* Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS

## 1.1.2 _(2019-12-22)_
* Fix: Use full path to CHANGELOG.md in the Changelog section of readme.txt
* Change: Note compatibility through WP 5.3+
* Change: Update copyright date (2020)

## 1.1.1 _(2019-06-11)_
* New: Add CHANGELOG.md file and move all but most recent changelog entries into it
* Change: Update unit test install script and bootstrap to use latest WP unit test repo
* Change: Note compatibility through WP 5.2+

## 1.1 _(2019-02-07)_
* Change: Initialize plugin on `plugins_loaded` action instead of on load
* Change: Merge `do_init()` into `init()`
* Change: Note compatibility through WP 5.1+
* Change: Add README.md link to plugin's page in Plugin Directory
* Change: Add more unit tests related to the registering of hooks
* Fix: Correct typo in GitHub URL.
* Change: Update copyright date (2019)
* Change: Update License URI to be HTTPS
* Change: Split paragraph in README.md's "Support" section into two

## 1.0 _(2018-01-22)_
* Initial public release


tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Comment_Link_Moderation_Whitelist
 */

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/comment-link-moderation-whitelist.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';



tests/test-comment-link-moderation-whitelist.php
<?php

defined( 'ABSPATH' ) or die();

class Comment_Link_Moderation_Whitelist_Test extends WP_UnitTestCase {

	protected static $setting_name = 'c2c-comment-link-moderation-whitelist';

	public function setUp() {
		parent::setUp();

		/** @var WP_REST_Server $wp_rest_server */
		global $wp_rest_server;
		$this->server = $wp_rest_server = new \WP_REST_Server;
		do_action( 'rest_api_init' );
	}

	public function tearDown() {
		parent::tearDown();
		$this->unset_current_user();
	}


	//
	// HELPER FUNCTIONS
	//


	private function create_user( $set_as_current = true ) {
		$user_id = $this->factory->user->create();
		if ( $set_as_current ) {
			wp_set_current_user( $user_id );
		}
		return $user_id;
	}

	// helper function, unsets current user globally. Taken from post.php test.
	private function unset_current_user() {
		global $current_user, $user_ID;

		$current_user = $user_ID = null;
	}

	private function check_comment( $comment_max_links = 2, $link_urls = array() ) {
		update_option( 'comment_whitelist', 0 );
		update_option( 'comment_max_links', $comment_max_links );

		$author       = 'BobtheBuilder';
		$author_email = 'bob@example.com';
		$author_url   = 'http://example.com';
		$author_ip    = '192.168.0.1';
		$user_agent   = '';
		$comment_type = '';

		if ( ! $link_urls ) {
			$comment  = 'This is a comment with <a href="http://example.com">multiple</a> <a href="http://bob.example.com">links</a>.';
		} else {
			$comment  = 'This is a comment with';
			foreach ( (array) $link_urls as $i => $link ) {
				// Prefix a protocol if one wasn't specified.
				if ( false === strpos( $link, 'http' ) ) {
					$link = 'https://' . $link;
				}
				$comment .= ' <a href="' . $link . '">link #' . $i . '</a>';
			}
			$comment .= '.';
		}

		return check_comment( $author, $author_email, $author_url, $comment, $author_ip, $user_agent, $comment_type );
	}


	//
	// TESTS
	//


	public function test_class_is_available() {
		$this->assertTrue( class_exists( 'c2c_CommentLinkModerationWhitelist' ) );
	}

	public function test_plugin_version() {
		$this->assertEquals( '1.1.3', c2c_CommentLinkModerationWhitelist::version() );
	}

	public function test_hooks_plugins_loaded() {
		$this->assertEquals( 10, has_action( 'plugins_loaded', array( 'c2c_CommentLinkModerationWhitelist', 'init' ) ) );
	}

	public function test_hooks_admin_init() {
		$this->assertEquals( 9, add_action( 'admin_init', array( 'c2c_CommentLinkModerationWhitelist', 'initialize_setting' ) ) );
	}

	public function test_hooks_comment_max_links_url() {
		$this->assertEquals( 10, has_filter( 'comment_max_links_url', array( 'c2c_CommentLinkModerationWhitelist', 'comment_max_links_url' ) ) );
	}

	// Duplicates core test of this behavior.
	public function test_should_return_false_when_link_count_exceeds_comment_max_length_setting() {
		$this->assertFalse( $this->check_comment() );
	}

	// Duplicates core test of this behavior.
	public function test_should_return_true_when_link_count_does_not_exceed_comment_max_length_setting() {
		$this->assertTrue( $this->check_comment( 3 ) );
	}

	public function test_link_count_not_exceeded_if_domain_is_whitelisted() {
		update_option( self::$setting_name, 'example.com' );

		$this->assertTrue( $this->check_comment() );
	}

	public function test_link_count_still_exceeded_if_different_domain_is_whitelisted() {
		update_option( self::$setting_name, 'example2.com' );

		$this->assertFalse( $this->check_comment() );
	}

	public function test_link_count_not_exceeded_if_whitelisted_domain_includes_path() {
		update_option( self::$setting_name, 'example.com/docs' );

		$this->assertTrue( $this->check_comment( 2, array( 'example.com', 'example.com/docs/doc-a', 'example.com/docs/doc-b' ) ) );
	}

	public function test_whitelisted_url_with_protocol_still_whitelisted() {
		update_option( self::$setting_name, 'https://example.com' );

		$this->assertTrue( $this->check_comment() );
	}

	public function test_whitelisted_url_still_whitelisted_when_one_of_many() {
		update_option( self::$setting_name, "nothing.com\nexample.com\nsomething.com" );

		$this->assertTrue( $this->check_comment() );
	}

	public function test_link_count_exceeded_by_links_to_same_domain_but_not_whitelisted_path() {
		update_option( self::$setting_name, 'example.com/docs' );

		$this->assertFalse( $this->check_comment( 2, array( 'example.com', 'example.com/info', 'example.com/docs' ) ) );
	}

	public function test_trailing_slash_in_whitelist_url_still_matches_url_without_slash() {
		update_option( self::$setting_name, 'example.com/' );

		$this->assertTrue( $this->check_comment() );
	}

	public function test_fallball_comment_max_links_limit_enforced_regardless_of_whitelist() {
		update_option( self::$setting_name, 'example.com' );

		$this->assertTrue( $this->check_comment( 2, array_fill( 0, 24, 'example.com' ) ) );
		$this->assertFalse( $this->check_comment( 2, array_fill( 0, 25, 'example.com' ) ) );
	}

	public function test_fallball_comment_max_links_limit_ignored_if_does_not_exceed_max_links_limit() {
		update_option( self::$setting_name, 'example.com' );

		$this->assertTrue( $this->check_comment( 26, array_fill( 0, 25, 'example.com' ) ) );
		$this->assertTrue( $this->check_comment( 50, array_fill( 0, 56, 'example.com' ) ) );
	}

	public function test_fallball_comment_max_links_limit_adjusted_to_exceed_large_max_links_limit() {
		update_option( self::$setting_name, 'example.com' );

		$this->assertTrue( $this->check_comment( 50, array_fill( 0, 59, 'example.com' ) ) );
		$this->assertFalse( $this->check_comment( 50, array_fill( 0, 60, 'example.com' ) ) );
	}

}


comment-link-moderation-whitelist.php
<?php
/**
 * Plugin Name: Comment Link Moderation Whitelist
 * Version:     1.1.3
 * Plugin URI:  https://coffee2code.com/wp-plugins/comment-link-moderation-whitelist/
 * Author:      Scott Reilly
 * Author URI:  https://coffee2code.com/
 * Text Domain: comment-link-moderation-whitelist
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Description: Allows for whitelisted URLs to not count against the comment moderation max number of links limit.
 *
 * Compatible with WordPress 4.7 through 5.4+.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/comment-link-moderation-whitelist/
 *
 * @package Comment_Link_Moderation_Whitelist
 * @author  Scott Reilly
 * @version 1.1.3
 */

/*
 * TODO:
 * - Support (via UI setting and/for filter and/or constant) the ability to define
 *   an alternative max links limit that takes into account raw link count (so the
 *   plugin can't be abused to post, say, 100 whitelisted URLs). Currently a
 *   hardcoded max limit is set (at 25) of the max number of whitelisted plus
 *   non-whitelisted links used in a comment.
 * - Add support for constant to disable admin UI and instead use whitelisted
 *   URLs defined via constnat.
 * - Add filter so list of whitelisted URLs can be programmatically modified.
 */

/*
	Copyright (c) 2018-2020 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_CommentLinkModerationWhitelist' ) ) :

class c2c_CommentLinkModerationWhitelist {

	/**
	 * Name for meta key used to store id of trashing user.
	 *
	 * @access private
	 * @var string
	 */
	private static $setting_name = 'c2c-comment-link-moderation-whitelist';

	/**
	 * Returns version of the plugin.
	 *
	 * @since 1.0
	 */
	public static function version() {
		return '1.1.3';
	}

	/**
	 * Hooks actions and filters.
	 *
	 * @since 1.0
	 */
	public static function init() {
		// Load textdomain.
		load_plugin_textdomain( 'comment-link-moderation-whitelist' );

		// Initialize setting.
		add_action( 'admin_init', array( __CLASS__, 'initialize_setting' ), 9 );

		// Register hooks.
		add_filter( 'comment_max_links_url', array( __CLASS__, 'comment_max_links_url' ), 10, 3 );
	}

	/**
	 * Initializes setting.
	 */
	public static function initialize_setting() {
		if ( ! current_user_can( 'manage_options' ) ) {
			return;
		}

		register_setting(
			'discussion',
			self::$setting_name,
			array(
				'type'              => 'string',
				'description'       => __( 'List of URLs that should not count towards the moderation max number of links.', 'comment-link-moderation-whitelist' ),
				'sanitize_callback' => array( __CLASS__, 'sanitize_option' ),
				'show_in_rest'      => false,
				'default'           => '',
			)
		);

		add_settings_field(
			self::$setting_name,
			__( 'Comment Link Moderation Whitelist', 'comment-link-moderation-whitelist' ),
			array( __CLASS__, 'display_option' ),
			'discussion',
			'default',
			array( 'label_for' => esc_attr( self::$setting_name ) )
		);
	}

	/**
	 * Sanitizes the option.
	 *
	 * Basically duplicates core's sanitization of 'moderation_keys' setting.
	 *
	 * @since 1.0
	 *
	 * @param string $value The value to sanitize.
	 * @return string
	 */
	public static function sanitize_option( $value ) {
		global $wpdb;

		$value = $wpdb->strip_invalid_text_for_column( $wpdb->options, 'option_value', $value );

		if ( is_wp_error( $value ) ) {
			$error = $value->get_error_message();
		} else {
			$value = explode( "\n", $value );
			$value = array_filter( array_map( 'trim', $value ) );
			$value = array_unique( $value );
			$value = implode( "\n", $value );
		}

		if ( ! empty( $error ) ) {
			$value = get_option( self::$setting_name );
			if ( function_exists( 'add_settings_error' ) ) {
				add_settings_error( self::$setting_name, 'invalid_' . self::$setting_name, $error );
			}
		}

		return $value;
	}

	/**
	 * Displays admin setting field.
	 *
	 * @since 1.0
	 *
	 * @param array $args Array of display arguments.
	 */
	public static function display_option( $args ) {
		$value = trim( get_option( self::$setting_name ) );

		echo '<fieldset>';
		echo '<legend class="screen-reader-text"><span>' . __( 'Comment Link Moderation Whitelist', 'comment-link-moderation-whitelist' ) . '</span></legend>';

		printf(
			'<p><label for="%s">%s</label></p>',
			esc_attr( self::$setting_name ),
			__( 'The URLs listed below will not count against the moderation link limit specified just above. One domain per line. Protocol (i.e. "http://") is not necessary. Subdomains are inferred and don\'t need to be listed individually (e.g. <em>wordpress.org</em> also implies <em>developer.wordpress.org</em>).', 'comment-link-moderation-whitelist' )
		);

		printf(
			'<p><textarea name="%s" rows="5" cols="50" id="%s" class="large-text code">%s</textarea></p>',
			esc_attr( self::$setting_name ),
			esc_attr( self::$setting_name ),
			esc_textarea( get_option( self::$setting_name ) )
		);

		// Add inline JS to move Comment Blacklist field to the bottom.
		echo "<script>try { jQuery('#blacklist_keys').closest('tr').appendTo(jQuery('#comment_moderation').closest('tbody')) } catch (err) {}</script>";

		echo "</fieldset>\n";
	}


	/**
	 * Adjusts the max number of links permitted in comments to disregard the
	 * number of whitelisted links.
	 *
	 * @since 1.0
	 *
	 * @param int    $num_links The number of links found.
	 * @param string $url       Comment author's URL. Included in allowed links total.
	 * @param string $comment   Content of the comment.
	 * @return int
	 */
	public static function comment_max_links_url( $num_links, $url, $comment ) {
		// Get comment max links.
		$max_links = get_option( 'comment_max_links' );

		// Bail if the number of links doesn't exceed the maximum.
		if ( $num_links < $max_links ) {
			return $num_links;
		}

		// Prevent abuse by enforcing a max link limit regardless of whitelist.
		$max_max_links = 25; // TODO: Make this a setting.

		// In the event site sets comment_max_links higher than max_max_limit,
		// adjust the fallback max to be 10 higher than comment_max_links.
		$max_max_links = ( $max_links > $max_max_links ) ? ( $max_links + 10 ) : $max_max_links;

		// Bail if the number of links exceeds the true max.
		if ( $num_links >= $max_max_links ) {
			return $num_links;
		}

		// Get whitelisted URLs.
		$whitelist_urls = trim( get_option( self::$setting_name ) );

		// Bail if no whitelisted URLs are defined.
		if ( ! $whitelist_urls ) {
			return $num_links;
		}

		// URLs are defined one per line.
		$whitelist_urls = explode( "\n", $whitelist_urls );

		// Check if any whitelisted URLs are present in comment.
		foreach ( $whitelist_urls as $url ) {

			// Filter out protocol.
			$url = str_replace( array( 'http://', 'https://' ), '', $url );

			// Remove trailing slash.
			$url = rtrim( $url, '/' );

			// Count the number of occurrences of this particular whitelisted URL.
			$num_whitelist_links = preg_match_all( '%<a [^>]*href=[\'\"]https?://[^/>]*' . preg_quote( $url, '%' ) . '%i', $comment, $out );

			// Increase the limit by the number of whitelisted URLs (so that they don't
			// count against the limit).
			$num_links -= $num_whitelist_links;

			// Stop if there are enough whitelisted links to bring the number of
			// non-whitelisted links below the max.
			if ( $num_links < $max_links ) {
				break;
			}

		}

		return $num_links;
	}

} // end c2c_CommentLinkModerationWhitelist

add_action( 'plugins_loaded', array( 'c2c_CommentLinkModerationWhitelist', 'init' ) );

endif; // end if !class_exists()
